[docs] Add API tradeoff section for the sx prop (#23962)
mui-org/material-ui: docs/src/pages/system/basics/basics.md

  We believe that for most uses it's **fast enough**, but there are simple workarounds when performance becomes critical. For instance, when rendering a list with many items, you can use a CSS child selector to have a single "style injection" point (using d. for the wrapper and a. for each item).

### API tradeoff

> In previous versions, the system properties were supported as props on the `Box` component. From v5, however, the system provides a superset of CSS (supports all CSS properties/selectors in addition to custom ones), and is available in all components,  so selectors cannot be efficiently mapped to props without potential naming conflicts. Instead, all system properties are available under one prop `sx`.

Reviewer: ```suggestion
In previous versions, the system properties were supported as props on the `Box` component. From v5, however, the system provides a superset of CSS (supports all CSS properties/selectors in addition to custom ones), and is available in all components, so selectors cannot be efficiently mapped to props without potential naming conflicts. Instead, all system properties are available under one prop `sx`.
```

> In previous versions, the system properties were supported as props on the `Box` component. From v5, however, the system provides a superset of CSS (supports all CSS properties/selectors in addition to custom ones), and is available in all components,  and so selectors cannot be efficiently mapped to props without potential naming conflicts. Instead, all system properties are available under one prop `sx`.

Author: ```suggestion
In previous versions, the system properties were supported as props on the `Box` component. From v5, however, the system provides a superset of CSS (supports all CSS properties/selectors in addition to custom ones), and is available in all components,  so selectors cannot be efficiently mapped to props without potential naming conflicts. Instead, all system properties are available under one prop `sx`.
```

  We believe that for most uses it's **fast enough**, but there are simple workarounds when performance becomes critical. For instance, when rendering a list with many items, you can use a CSS child selector to have a single "style injection" point (using d. for the wrapper and a. for each item).

### API tradeoff

> In the previous versions, we supported the system properties as props on the `Box` component. However, the system is a superset of CSS (contains all CSS properties/selectors in addition to custom ones) and selectors cannot be efficiently mapped to props without potential conflicts. This is the main reason why, we decided to support all system properties under one prop - `sx`, which in addition allows developers to define other CSS definitions.

Reviewer 2: ```suggestion
In previous versions, the system properties were supported as props on the `Box` component. From v5, however, the system provides a superset of CSS (supports all CSS properties/selectors in addition to custom ones), and is available in all components,  and so selectors cannot be efficiently mapped to props without potential naming conflicts. Instead, all system properties are available under one prop `sx`.
```

> In the previous versions, we supported the system properties as first class properties on the `Box` component. However, as we are extending the system to be a superset of CSS (contains all CSS properties/selectors in addition to custom ones) it's no longer an option to support all properties as first class props. This is the main reason why, we decided to support all system properties under one prop - `sx`, which in addition allows developers to define other CSS definitions.

Reviewer: Matt might suggest to replace "we", not sure

```suggestion
In the previous versions, we supported the system properties as first-class properties on the `Box` component. However, as we are extending the system to be a superset of CSS (contains all CSS properties/selectors in addition to custom ones) it's no longer an option to support all properties as first-class props. This is the main reason why, we decided to support all system properties under one prop - `sx`, which in addition allows developers to define other CSS definitions.
```

Reviewer 3: > Matt might suggest to replace "we", not sure 

I think I forget this one as well. @mbrookes explained it once but I kind of forgot :blush:

Author: I cannot get over this :D

Reviewer 3: Can't be good at everything :wink: There's a reason technical writers exist.

> In the previous versions, we supported the system properties as first class properties on the `Box` component. However, as we are extending the system to be a superset of CSS (contains all CSS properties/selectors in addition to custom ones) it's no longer an option to support all properties as first class props. This is the main reason why, we decided to support all system properties under one prop - `sx`, which in addition allows developers to define other CSS definitions.

Reviewer 3: ```suggestion
In the previous versions, we supported the system properties as props on the `Box` component. However, the system is a superset of CSS and selectors cannot be efficiently mapped to props without potential conflicts. This is the main reason why, we decided to support all system properties under one prop - `sx`, which in addition allows developers to define other CSS definitions.
```

1. I don't think there's a notion of "first-class props", just props
1. As far as I can tell it's not about mapping system properties to props but rather that we can't efficiently map selectors. So we consolidate it under a single prop to avoid refactoring hazard when going from "just system properties" to system properties + selectors. Kind of like how inline styles should be avoided because you'll eventually have to refactor to style rules.

Author: Agreed, and even more, is not just about the selectors, the there are many other CSS Properties, that are not part of the system, which we would need to map to props...

Author: I still think we should describe what superset of CSS means, as it seems like it is confusing for people. What do you think of this - https://github.com/mui-org/material-ui/pull/23962/files#r541019182 as final?

Author: I've applied https://github.com/mui-org/material-ui/pull/23962/files#r541019182 , feel free to leave suggestion if anything else needs to be updated

> In the previous versions, we supported the system properties as first-class properties on the `Box` component. However, as we are extending the system to be a superset of CSS (contains all CSS properties/selectors in addition to custom ones) it's no longer an option to support all properties as first-class props. This is the main reason why, we decided to support all system properties under one prop - `sx`, which in addition allows developers to define other CSS definitions.

Author: ```suggestion
In the previous versions, we supported the system properties as props on the `Box` component. However, the system is a superset of CSS (contains all CSS properties/selectors in addition to custom ones) and selectors cannot be efficiently mapped to props without potential conflicts. This is the main reason why, we decided to support all system properties under one prop - `sx`, which in addition allows developers to define other CSS definitions.
```

> In order to support the system on all core components, having the system under one prop, allows avoiding collision with the props, that may already exist on the components, like `color`.

Reviewer 2: ```suggestion
```
NOw covered above.

> 

Reviewer 2: ```suggestion
```

> As we are working on supporting the `sx` on all core components, having the system under one prop, helps us in order to avoid collision with the properties, that may already exists on the components, like `color`.

Reviewer: ```suggestion
In order to support the system on all core components, having the system under one prop, allows avoiding collision with the props, that may already exist on the components, like `color`.
```

Reviewer 3: > Matt might suggest to replace "we", not sure 

I think I forget this one as well. @mbrookes explained it once but I kind of forgot :blush:

> Finally, having the system under one prop helps to easily differentiate the props defined for the sole purpose of CSS utilities vs. for component business logic.

Reviewer 2: ```suggestion
Additionally, having the system under one prop helps to easily differentiate props defined for the sole purpose of CSS utilities, vs. those for component business logic.
```

> Finally, having the props under one property, helps to easily differentiate the component's prop versus the props defined for the sole purpose of styling.

Reviewer: ```suggestion
Finally, having the system under one prop helps to easily differentiate the props defined for the sole purpose of CSS utilities vs. for component business logic.
```
